src/client/js/router.js

```diff
@@ -0,0 +1,151 @@
+/**
+ *
+ *  Online store PWA sample.
+ *  Copyright 2017 Google Inc. All rights reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License
+ *
+ */
+
+import { instance as lazyImg } from './lazy-img.js';
+import highlightCategory from './highlight-category.js';
+
+class Router {
+  constructor() {
+    this._bindHandlers();
+    this._hostname = location.host;
+
+    document.addEventListener('click', this._onLinkClick);
+    window.addEventListener('popstate', this._onPopState);
+  }
+
+  _bindHandlers() {
+    this._onLinkClick = this._onLinkClick.bind(this);
+    this._onPopState = this._onPopState.bind(this);
+  }
+
+  _getParentByTagName(node, tagname) {
+    let parent;
+    if (node === null || tagname === '') return;
+    parent = node.parentNode;
+    tagname = tagname.toUpperCase();
+
+    while (parent.tagName !== "HTML") {
+      if (parent.tagName === tagname) {
+        return parent;
+      }
+      parent = parent.parentNode;
+    }
+    return parent;
+  }
+
+  _onPopState(event) {
+    this.navigate(new URL(location.href), event.state.scrollTop, false);
+  }
+
+  async _onLinkClick(event) {
+    let target = event.target;
+
+    if (target.tagName !== 'A') {
+      target = this._getParentByTagName(target, 'A');
+      if (!target) return;
+    }
+    const link = new URL(target.href);
```

Comment:
I get this error when I click on the menu icon: 

`Uncaught (in promise) TypeError: Failed to construct 'URL': Invalid URL
    at Router._onLinkClick (router.js:63)`

Reply:
Added additional check